include/qt_atomics.h

```diff
@@ -54,9 +67,12 @@ void qt_spin_exclusive_unlock(qt_spin_exclusive_t *);
 # define QTHREAD_FASTLOCK_INIT(x)     { (x).enter = 0; (x).exit = 0; }
 # define QTHREAD_FASTLOCK_INIT_PTR(x) { (x)->enter = 0; (x)->exit = 0; }
 # define QTHREAD_FASTLOCK_LOCK(x)     { aligned_t val = qthread_incr(&(x)->enter, 1); \
-                                        while (val != (x)->exit) SPINLOCK_BODY(); /* spin waiting for my turn */ }
+                                        while (val != (x)->exit) SPINLOCK_BODY(); \
+                                          THREAD_FENCE_MEM_ACQUIRE; /* spin waiting for my turn */ }
 # define QTHREAD_FASTLOCK_UNLOCK(x)   do { COMPILER_FENCE; \
-                                           (x)->exit++; /* allow next guy's turn */ } while (0)
+                                           (x)->exit++; /* allow next guy's turn */ \
+                                           THREAD_FENCE_MEM_RELEASE; \
```

Comment:
Doesn't the release fence need to be before the code that actually releases the lock? As-is it seems like a thread could execute `(x)->exit++;`, get interrupted and have another thread run and acquire the lock before any pending operations from the original thread are fenced.

Reply:
Yes.

Reply:
Yes, seem wrong here. Even without interruption. Another thread can be polling the while-loop on trying to enter while the thread that has entered the critical section is updating the exit++, thus allowing the polling thread to enter without the exiting thread to call `THREAD_FENCE_MEM_RELEASE`.

Reply:
FYI changing the order allowed 1K trials of my test to pass.

Reply:
Cool, thanks.